Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -214,71 +354,109 @@ static RefPtr<SharedBuffer> parseResponseMessage(const String& message)
+    OpenCDMSession* session = nullptr;
+    opencdm_create_session(&source, keySystem.utf8().data(), openCDMLicenseType(licenseType), mimeType, reinterpret_cast<const uint8_t*>(m_initData->data()), m_initData->size(),
+        !customData->isEmpty() ? reinterpret_cast<const uint8_t*>(customData->data()) : nullptr, customData->size(), callbacksNotifier.callbacks(), &session);
+    if (!session)
         return;
+    m_session = ScopedSession(session, [](OpenCDMSession* session) {
+        opencdm_session_close(session);
+        opencdm_destruct_session(session);
+    });
+    m_id = String::fromUTF8(opencdm_session_id(m_session.get()));
+    callbacksNotifier.addObserver(m_session.get(), this);
+}
 
-    m_isValid = temporaryURLLength;
+CDMInstanceOpenCDM::Session::~Session()
+{
+    OCDMSessionCallbacksNotifier::singleton().removeObserver(this);
+}
 
+void CDMInstanceOpenCDM::Session::challengeGeneratedCallback(String&& message)
+{
     // We could do all operations with String but this way we can save some copies.
     size_t typePosition = message.find(":Type:");
-    String requestType(message.c_str(), typePosition != std::string::npos ? typePosition : 0);
+    String requestType(message.characters8(), typePosition != notFound ? typePosition : 0);
     unsigned offset = 0;
-    if (!requestType.isEmpty() && requestType.length() != message.size())
+    if (!requestType.isEmpty() && requestType.length() != message.length())
         offset = typePosition + 6;
 
-    m_message = SharedBuffer::create(message.c_str() + offset, message.size() - offset);
+    m_message = SharedBuffer::create(message.characters8() + offset, message.length() - offset);
     m_needsIndividualization = requestType.length() == 1 && static_cast<WebCore::MediaKeyMessageType>(requestType.toInt()) == CDMInstance::MessageType::IndividualizationRequest;
+    for (const auto& challengeCallback : m_challengeCallbacks)
+        challengeCallback(this);
+    m_challengeCallbacks.clear();
 }
 
-std::pair<media::OpenCdm::KeyStatus, String> CDMInstanceOpenCDM::Session::update(const uint8_t* data, const unsigned length)
+void CDMInstanceOpenCDM::Session::keyUpdatedCallback(String&& status)
 {
-    std::string response;
-    m_lastStatus = m_session.Update(data, length, response);
-    return std::make_pair(m_lastStatus, String(response.c_str(), response.empty() ? 0 : response.size()));
+    for (auto& pendingKeyUpdate : m_pendingKeyUpdates) {
+        m_lastStatus = opencdm_session_status(m_session.get(), pendingKeyUpdate.m_key.get(), pendingKeyUpdate.m_keySize);
+        pendingKeyUpdate.m_callback(this, m_lastStatus, status);
+    }
+    m_pendingKeyUpdates.clear();
+
+    for (auto& sessionChangedCallback : m_sessionChangedCallbacks)
+        sessionChangedCallback(this, true, status);
+    m_sessionChangedCallbacks.clear();
 }
 
-std::pair<bool, String> CDMInstanceOpenCDM::Session::load()
+void CDMInstanceOpenCDM::Session::messageReceivedCallback(String&& message)
```

Comment:
Why is this empty?

Reply:
IT doesn't seem needed. I do not think there is anything to do with the message yet the callback needs to be registered.